// @zohodesk/testinglibrary/build/utils/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Logger = void 0;
// const SUCCESS_TYPE = 'success';
// const FAILURE_TYPE = 'failure';
// const INFO_TYPE = 'info';

// function logger() {
//   this.colors = {
//     'success': ['\x1b[36m', '\x1b[0m'],
//     'failure': ['\x1b[31m', '\x1b[0m'],
//     'info': ['\x1b[33m', '\x1b[0m']
//   }
//   this.consoleLogger = console;
//   return {
//     SUCCESS_TYPE,
//     FAILURE_TYPE,
//     INFO_TYPE,
//     error: () => { },
//     info: () => { },
//     log: (type, message) => {
//       const color = this.colors[type];
//       console.log(type, color)
//       this.consoleLogger.log(`${color[0]}${message}${color[1]}`)
//     }
//   }
// }

// module.exports = {
//   Logger: logger()
// }

class LoggerImpl {
  constructor() {
    this.SUCCESS_TYPE = 'success';
    this.FAILURE_TYPE = 'failure';
    this.INFO_TYPE = 'info';
    this.colors = {
      'success': ['\x1b[36m', '\x1b[0m'],
      'failure': ['\x1b[31m', '\x1b[0m'],
      'info': ['\x1b[33m', '\x1b[0m']
    };
    this.consoleLogger = console;
  }
  error(err) {
    this.consoleLogger.error(err);
  }
  info() {}
  log(type, message) {
    const color = this.colors[type];
    this.consoleLogger.log(`${color[0]}${message}${color[1]}`);
  }
}
const Logger = exports.Logger = new LoggerImpl();
