// @zohodesk/testinglibrary/build/setup-folder-structure/samples/uat-config-sample.js
const testSetup = require('../../fixtures/testSetup');
/**
 * @typedef {Object|null} viewportConfig
 * @property {number} width - width of the viewport
 * @property {number} height - height of the viewport 
 */
/**
 * Represents the user configuration object.
 * @typedef {Object} UserConfig
 * @property {string} headless - Headless Browsers mode.
 * @property {number} trace - trace for test cases.
 * @property {boolean} video - video for test cases,
 * @property {boolean} debug - debug mode
 * @property {string} mode: mode in which the test cases needs to run
 * @property {boolean} isAuthMode -  Auth Mode. config whether authentication step needed before running test cases
 * @property {string} authFilePath - File Path where the cookies stored
 * @property {any} browsers: List of browsers
 * @property {string} openReportOn: default Option value (never, on-failure and always) 
 * @property {any} reportPath : directory where report is generate
 * @property {boolean} bddMode: Feature files needs to be processed
 * @property {number} expectTimeout: time in milliseconds which the expect condition should fail
 * @property {number} testTimeout: time in milliseconds which the test should fail
 * @property {Object} additionalPages: custom pages configuration
 * @property {string} featureFilesFolder: folder name under which feature-files will be placed. Default is feature-files
 * @property {string} stepDefinitionsFolder: folder name under which step implementations will be placed. Default is steps
 * @property {viewportConfig} viewport: viewport configuration for the browser. Default is { width: 1280, height: 720 }
 * @property {string} testIdAttribute: Change the default data-testid attribute. configure what attribute to search while calling getByTestId
 */

/**
* @type {UserConfig}
*/
module.exports = {
  headless: false,
  browsers: ['Chrome', 'Firefox', 'Safari', 'Edge'],
  mode: 'dev',
  isAuthMode: true,
  authFilePath: 'uat/playwright/.auth/user.json',
  trace: true,
  video: true,
  bddMode: true,
  featureFilesFolder: 'feature-files',
  stepDefinitionsFolder: 'steps',
  viewport: { width: 1280, height: 720 },
  testSetup
}
