// @zohodesk/testinglibrary/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.When = exports.Then = exports.Step = exports.Given = exports.But = exports.And = void 0;
Object.defineProperty(exports, "accountLogin", {
  enumerable: true,
  get: function () {
    return _index.accountLogin;
  }
});
Object.defineProperty(exports, "createBdd", {
  enumerable: true,
  get: function () {
    return _index.createBdd;
  }
});
Object.defineProperty(exports, "expect", {
  enumerable: true,
  get: function () {
    return _index.expect;
  }
});
Object.defineProperty(exports, "getDefaultActor", {
  enumerable: true,
  get: function () {
    return _index.getDefaultActor;
  }
});
Object.defineProperty(exports, "getListOfActors", {
  enumerable: true,
  get: function () {
    return _index.getListOfActors;
  }
});
Object.defineProperty(exports, "getRunMode", {
  enumerable: true,
  get: function () {
    return _index.getRunMode;
  }
});
Object.defineProperty(exports, "getUserForSelectedEditionAndProfile", {
  enumerable: true,
  get: function () {
    return _index.getUserForSelectedEditionAndProfile;
  }
});
Object.defineProperty(exports, "isCI", {
  enumerable: true,
  get: function () {
    return _index.isCI;
  }
});
Object.defineProperty(exports, "isDevelopmentSetup", {
  enumerable: true,
  get: function () {
    return _index.isDevelopmentSetup;
  }
});
Object.defineProperty(exports, "loadCookiesIfPresent", {
  enumerable: true,
  get: function () {
    return _index.loadCookiesIfPresent;
  }
});
Object.defineProperty(exports, "performLoginSteps", {
  enumerable: true,
  get: function () {
    return _index.performLoginSteps;
  }
});
Object.defineProperty(exports, "setup", {
  enumerable: true,
  get: function () {
    return _index.setup;
  }
});
Object.defineProperty(exports, "test", {
  enumerable: true,
  get: function () {
    return _index.test;
  }
});
Object.defineProperty(exports, "verifyIfCookieFileExists", {
  enumerable: true,
  get: function () {
    return _index.verifyIfCookieFileExists;
  }
});
var _index = require("./core/playwright/index");
// import { fireEvent, render } from '@testing-library/react';

const {
  Given,
  Then,
  When,
  Step
} = (0, _index.createBdd)();
exports.Step = Step;
exports.When = When;
exports.Then = Then;
exports.Given = Given;
const And = exports.And = Then;
const But = exports.But = Then;
