// @zohodesk/testinglibrary/playwright.config.js
// @ts-check
const { defineConfig, devices } = require('@playwright/test');
const path = require('path');
const numCPUs = require('os').cpus().length;

export default defineConfig({
  testDir: path.join(path.resolve(process.cwd()), 'uat'),
  outputDir: path.join(process.cwd(), 'uat', 'test-results'),
  fullyParallel: true,
  retries: process.env.CI ? 2 : 0,
  reporter: [['html', { outputFolder: path.join(process.cwd(), 'uat', 'playwright-report'), open: "always" }]],
  timeout: 60 * 1000,
  expect: {
    timeout: 5 * 1000,
  },
  use: {
    trace: 'on',
    video: {
      mode: 'on',
      size: { width: 640, height: 480 }
    }
  },
  projects: [
    { name: 'setup', testMatch: /.*\.setup\.js/ },
    {
      name: 'chromium',
      use: {
        ...devices['Desktop Chrome'],
        storageState: path.resolve(process.cwd(), 'uat', 'playwright/.auth/user.json')
      },
      dependencies: ['setup'],
    },

    {
      name: 'firefox',
      timeout: 4 * 60 * 1000,
      expect: {
        timeout: 80 * 1000,
      },
      use: {
        ...devices['Desktop Firefox'],
        storageState: path.resolve(process.cwd(), 'uat', 'playwright/.auth/user.json')
      },
      dependencies: ['setup'],
    },

    {
      name: 'webkit',
      timeout: 2 * 60 * 1000,
      expect: {
        timeout: 80 * 1000,
      },
      use: {
        ...devices['Desktop Safari'],
        storageState: path.resolve(process.cwd(), 'uat', 'playwright/.auth/user.json')
      },
      dependencies: ['setup'],
    },

  ],
});
